service/share/backoff.go

```diff
@@ -0,0 +1,99 @@
+package share
+
+import (
+	"context"
+	"fmt"
+	"sync"
+	"time"
+
+	"github.com/libp2p/go-libp2p-core/host"
+	"github.com/libp2p/go-libp2p-core/peer"
+	"github.com/libp2p/go-libp2p/p2p/discovery/backoff"
+)
+
+// defaultTimeToLive is a default period after which disconnected peers will be removed from cache
+const defaultTimeToLive = time.Hour
+
+var defaultBackoffFactory = backoff.NewFixedBackoff(time.Hour)
+
+// backoffConnector wraps a libp2p.Host to establish a connection with peers
+// with adding a delay for the next connection attempt.
+type backoffConnector struct {
+	h       host.Host
+	backoff backoff.BackoffFactory
+
+	cacheLk   sync.Mutex
+	cacheData map[peer.ID]*connectionCacheData
+}
+
+// connectionCacheData stores time when next connection attempt with the remote peer.
+type connectionCacheData struct {
+	nexttry time.Time
+	backoff backoff.BackoffStrategy
+}
+
+func newBackoffConnector(h host.Host, factory backoff.BackoffFactory) *backoffConnector {
+	return &backoffConnector{
+		h:         h,
+		backoff:   factory,
+		cacheData: make(map[peer.ID]*connectionCacheData),
+	}
+}
+
+// Connect puts peer to the backoffCache and tries to establish a connection with it.
+func (b *backoffConnector) Connect(ctx context.Context, p peer.AddrInfo) error {
+	strategy := b.backoff()
+	b.cacheLk.Lock()
+	cacheData, ok := b.cacheData[p.ID]
+	if ok {
+		if time.Now().Before(cacheData.nexttry) {
+			b.cacheLk.Unlock()
+			return fmt.Errorf("share/discovery: backoff period is not ended for peer=%s", p.ID.String())
+		}
+		strategy = b.cacheData[p.ID].backoff
+	}
+	cacheData = &connectionCacheData{
+		nexttry: time.Now().Add(strategy.Delay()),
+		backoff: strategy,
+	}
+	b.cacheData[p.ID] = cacheData
```

Comment:
Avoid reallocating the cachedata here for the case where it is already on the map.

Reply:
Not fixed yet